pkg/kubelet/status_manager.go

```diff
@@ -104,7 +104,22 @@ func (s *statusManager) SyncBatch() {
 			podFullName := kubecontainer.GetPodFullName(pod)
 			status := syncRequest.status
 			glog.V(3).Infof("Syncing status for %s", podFullName)
-			_, err := s.kubeClient.Pods(pod.Namespace).UpdateStatus(pod.Name, &status)
+
+			var err error
+			statusPod := &api.Pod{
+				ObjectMeta: pod.ObjectMeta,
+			}
+			// TODO: make me easier to express from client code
+			// TODO: resource version on a pod should be saved when dealing with mirror pods so that
+			// subsequent updates don't require a GET.
```

Comment:
----- Original Message -----
> > @@ -104,7 +104,22 @@ func (s *statusManager) SyncBatch() {
> >  			podFullName := kubecontainer.GetPodFullName(pod)
> >  			status := syncRequest.status
> >  			glog.V(3).Infof("Syncing status for %s", podFullName)
> > -			_, err := s.kubeClient.Pods(pod.Namespace).UpdateStatus(pod.Name,
> > &status)
> > +
> > +			var err error
> > +			statusPod := &api.Pod{
> > +				ObjectMeta: pod.ObjectMeta,
> > +			}
> > +			// TODO: make me easier to express from client code
> > +			// TODO: resource version on a pod should be saved when dealing with
> > mirror pods so that
> > +			// subsequent updates don't require a GET.
> 
> I don't quite understand what makes mirror pods different from regular pods
> in this case...

Regular pods are getting updates from other sources, so the Kubelet's watch on the master will report changes to them and fill an updated resource version, but it might lag (so the resource version the kubelet knows is less likely to be the most recent version).  Mirror pods start from the Kubelet, so the resource version of a mirror pod update is known almost all the time.  It's not a huge distinction.

In general, the resource version the kubelet has last seen should be used instead of doing the GET here.  Done right, most status updates would be a single operation (a PUT) without conflict.  However, if there is a race, the kubelet needs to retry the status update.  We've discussed letting the server allow a server-overwrite for certain actions, which would put the logic on the server side, but that isn't implemented.  In general, avoiding the GET would be ideal, but I'm not convinced the Kubelet will be the only client that is setting status either (network management plugins?).  So we will eventually need to intelligently merge the status on the Kubelet side during an update.

> 
> Also, do we always want to overwrite the status even though kubelet was
> syncing to an older pod spec?

I'm sure there's some unpleasant race conditions and timing problems here.  The Kubelet may wish to delay updating status when its pod spec doesn't match the master's.  Also, there is a general problem that a 409 should return the resource as it exists on the server so that a client doesn't need to re-get.  That would reduce the complexity for clients that wish to implement overwrite or atomic retry.  I have a todo to open an issue on it.

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/5738/files#r27162095